Reading from Parquet file generates an assertion error

I'm guessing something changed between Spark 2.2 and 2.3 that we didn't catch because the unit tests around Parquet IO in `EncodingSpec` only counted rows, and didn't perform any action that realized the tiles.

```
java.lang.AssertionError: assertion failed: User-defined types in Catalyst schema should have already been expanded:

{
  "type" : "struct",
  "fields" : [ {
    "name" : "tile",
    "type" : {
      "type" : "struct",
      "fields" : [ {
        "name" : "cell_context",
        "type" : {
          "type" : "struct",
          "fields" : [ {
            "name" : "cellType",
            "type" : {
              "type" : "struct",
              "fields" : [ {
                "name" : "cellTypeName",
                "type" : "string",
                "nullable" : false,
                "metadata" : { }
              } ]
            },
            "nullable" : false,
            "metadata" : { }
          }, {
            "name" : "dimensions",
            "type" : {
              "type" : "struct",
              "fields" : [ {
                "name" : "cols",
                "type" : "short",
                "nullable" : false,
                "metadata" : { }
              }, {
                "name" : "rows",
                "type" : "short",
                "nullable" : false,
                "metadata" : { }
              } ]
            },
            "nullable" : false,
            "metadata" : { }
          } ]
        },
        "nullable" : false,
        "metadata" : { }
      }, {
        "name" : "cell_data",
        "type" : {
          "type" : "struct",
          "fields" : [ {
            "name" : "cells",
            "type" : "binary",
            "nullable" : true,
            "metadata" : { }
          }, {
            "name" : "ref",
            "type" : {
              "type" : "struct",
              "fields" : [ {
                "name" : "source",
                "type" : {
                  "type" : "udt",
                  "class" : "org.apache.spark.sql.rf.RasterSourceUDT",
                  "pyClass" : "pyrasterframes.rf_types.RasterSourceUDT",
                  "sqlType" : {
                    "type" : "struct",
                    "fields" : [ {
                      "name" : "raster_source_kryo",
                      "type" : "binary",
                      "nullable" : false,
                      "metadata" : { }
                    } ]
                  }
                },
                "nullable" : false,
                "metadata" : { }
              }, {
                "name" : "bandIndex",
                "type" : "integer",
                "nullable" : false,
                "metadata" : { }
              }, {
                "name" : "subextent",
                "type" : {
                  "type" : "struct",
                  "fields" : [ {
                    "name" : "xmin",
                    "type" : "double",
                    "nullable" : false,
                    "metadata" : { }
                  }, {
                    "name" : "ymin",
                    "type" : "double",
                    "nullable" : false,
                    "metadata" : { }
                  }, {
                    "name" : "xmax",
                    "type" : "double",
                    "nullable" : false,
                    "metadata" : { }
                  }, {
                    "name" : "ymax",
                    "type" : "double",
                    "nullable" : false,
                    "metadata" : { }
                  } ]
                },
                "nullable" : true,
                "metadata" : { }
              }, {
                "name" : "subgrid",
                "type" : {
                  "type" : "struct",
                  "fields" : [ {
                    "name" : "colMin",
                    "type" : "integer",
                    "nullable" : false,
                    "metadata" : { }
                  }, {
                    "name" : "rowMin",
                    "type" : "integer",
                    "nullable" : false,
                    "metadata" : { }
                  }, {
                    "name" : "colMax",
                    "type" : "integer",
                    "nullable" : false,
                    "metadata" : { }
                  }, {
                    "name" : "rowMax",
                    "type" : "integer",
                    "nullable" : false,
                    "metadata" : { }
                  } ]
                },
                "nullable" : true,
                "metadata" : { }
              } ]
            },
            "nullable" : true,
            "metadata" : { }
          } ]
        },
        "nullable" : false,
        "metadata" : { }
      } ]
    },
    "nullable" : true,
    "metadata" : { }
  } ]
}

```

--- a/core/src/main/scala/org/locationtech/rasterframes/ref/RasterRef.scala
+++ b/core/src/main/scala/org/locationtech/rasterframes/ref/RasterRef.scala
@@ -77,7 +77,7 @@ object RasterRef extends LazyLogging {
   implicit val rasterRefSerializer: CatalystSerializer[RasterRef] = new CatalystSerializer[RasterRef] {
     val rsType = new RasterSourceUDT()
     override def schema: StructType = StructType(Seq(
-      StructField("source", rsType, false),
+      StructField("source", rsType.sqlType, false),
       StructField("bandIndex", IntegerType, false),
       StructField("subextent", schemaOf[Extent], true),
       StructField("subgrid", schemaOf[GridBounds], true)

--- a/core/src/test/scala/org/locationtech/rasterframes/TestEnvironment.scala
+++ b/core/src/test/scala/org/locationtech/rasterframes/TestEnvironment.scala
@@ -62,8 +62,10 @@ trait TestEnvironment extends FunSpec with GeoTrellisTestEnvironment
     val dest = Files.createTempFile(Paths.get(outputLocalPath), "rf", ".parquet")
     logger.trace(s"Writing '${sanitized.columns.mkString(", ")}' to '$dest'...")
     sanitized.write.mode(SaveMode.Overwrite).parquet(dest.toString)
-    val rows = df.sparkSession.read.parquet(dest.toString).collect()
-    logger.trace(s" read back $rows row(s)")
+    val in = df.sparkSession.read.parquet(dest.toString)
+    // NB: The `collect` ensures values get fully reified.
+    val rows = in.collect()
+    logger.trace(s" read back ${rows.length} row(s)")
     rows.length == inRows
   }
 

--- a/core/src/test/scala/org/locationtech/rasterframes/TestEnvironment.scala
+++ b/core/src/test/scala/org/locationtech/rasterframes/TestEnvironment.scala
@@ -62,9 +62,9 @@ trait TestEnvironment extends FunSpec with GeoTrellisTestEnvironment
     val dest = Files.createTempFile(Paths.get(outputLocalPath), "rf", ".parquet")
     logger.trace(s"Writing '${sanitized.columns.mkString(", ")}' to '$dest'...")
     sanitized.write.mode(SaveMode.Overwrite).parquet(dest.toString)
-    val rows = df.sparkSession.read.parquet(dest.toString).count()
+    val rows = df.sparkSession.read.parquet(dest.toString).collect()
     logger.trace(s" read back $rows row(s)")
-    rows == inRows
+    rows.length == inRows
   }
 
   /**

--- a/docs/src/main/paradox/release-notes.md
+++ b/docs/src/main/paradox/release-notes.md
@@ -1,10 +1,12 @@
-# Release&nbsp;Notes
+# Release Notes
 
 ## 0.8.x
 
 ### 0.8.2
 
-* Fixed `TileExploder` to support `proj_raster` struct [(#287)](https://github.com/locationtech/rasterframes/issues/287).
+* Fixed Parquet serialization issue with `RasterRef`s ([#338](https://github.com/locationtech/rasterframes/issues/338))
+* Fixed `TileExploder`, `rf_agg_local_mean` and `TileColumnSupport` to support `proj_raster` struct ([#287](https://github.com/locationtech/rasterframes/issues/287), [#163](https://github.com/locationtech/rasterframes/issues/163), [#333](https://github.com/locationtech/rasterframes/issues/333)).
+* Various documentation improvements.
 
 ### 0.8.1
 

